ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -0,0 +1,822 @@
+  blockPreHeader = unAnnotated . CC.Block.recoverSignedBytes given
+                   . unByronHeader
+
+-- TODO get rid of this once we have a BlockHeader type family
+instance (Given CC.Slot.EpochSlots, Given CC.Block.HeaderHash, Typeable cfg) => HasPayload (PBft PBftCardanoCrypto) (ByronHeader cfg) where
+  blockPayload _ (ByronHeader header) = PBftPayload
+    { pbftIssuer = VerKeyCardanoDSIGN
+                   . Crypto.pskDelegateVK
+                   . Crypto.psigPsk
+                   . CC.Block.unBlockSignature
+                   . CC.Block.headerSignature
+                   $ header
+    , pbftSignature = SignedDSIGN
+                      . SigCardanoDSIGN
+                      . Crypto.Signature
+                      . Crypto.psigSig
+                      . CC.Block.unBlockSignature
+                      . CC.Block.headerSignature
+                      $ header
+    }
+
+
+instance (Given CC.Slot.EpochSlots, Given CC.Block.HeaderHash, Typeable cfg) => HasPayload (PBft PBftCardanoCrypto) (ByronBlock cfg) where
+  blockPayload cfg = blockPayload cfg . byronHeader
+
+-- | Override the delegation map from the ledger view
+--
+-- This is to work around a bug in cardano-ledger
+-- <https://github.com/input-output-hk/cardano-ledger/issues/504>
+reconstructDelegationMap :: Bimap CC.Common.StakeholderId CC.Common.StakeholderId
+reconstructDelegationMap =
+    go $ Genesis.gdHeavyDelegation Dummy.dummyGenesisData
+  where
+    go :: Genesis.GenesisDelegation
+       -> Bimap CC.Common.StakeholderId CC.Common.StakeholderId
+    go = Bimap.fromList . map go' . Map.toList . Genesis.unGenesisDelegation
+
+    go' :: (CC.Common.StakeholderId, Delegation.Certificate)
+        -> (CC.Common.StakeholderId, CC.Common.StakeholderId)
+    go' (from, to) =
+        if issuer /= from
+          then error "reconstructDelegationMap: unexpected issuer"
+          else (from, delegate)
+      where
+        issuer, delegate :: CC.Common.StakeholderId
+        issuer   = CC.Common.mkStakeholderId $ Crypto.pskIssuerVK   to
+        delegate = CC.Common.mkStakeholderId $ Crypto.pskDelegateVK to
+
+instance ( Given Crypto.ProtocolMagicId
+         , Given CC.Slot.EpochSlots
+         , Given CC.Block.HeaderHash
+         , Typeable cfg
+         ) => ProtocolLedgerView (ByronBlock cfg) where
+  protocolLedgerView _ns (ByronLedgerState ls _) = PBftLedgerView
+    -- Delegation map
+    ( Delegation.unMap
+      . V.Interface.delegationMap
+      . CC.Block.cvsDelegationState
+      $ ls
+    )
+
+  -- There are two cases here:
+  --
+  -- - The view we want is in the past. In this case, we attempt to find a
+  --   snapshot which contains the relevant slot, and extract the delegation map
+  --   from that.
+  --
+  -- - The view we want is in the future. In this case, we need to check the
+  --   upcoming delegations to see what new delegations will be made in the
+  --   future, and update the current delegation map based on that.
```

Comment:
Perhaps we should create a ticket for another thorough review and cleanup of this module once @nc6 is back.